File: server/src/main/java/com/db/datastoreserver/service/PostService.java
package com.db.datastoreserver.service;

import com.db.datastoreserver.domain.member.Member;
import com.db.datastoreserver.domain.post.Category;
import com.db.datastoreserver.domain.post.Photo;
import com.db.datastoreserver.domain.post.Post;
import com.db.datastoreserver.domain.post.repository.PostRepository;
import com.db.datastoreserver.service.dto.PostCreateRequest;
import com.db.datastoreserver.service.dto.PostResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

@RequiredArgsConstructor
@Service
public class PostService {

    private final PostRepository postRepository;
    private final UploadService uploadService;

    public List<PostResponse> findAllPosts() {
        return PostResponse.of(postRepository.findAll());
    }

    public List<PostResponse> findAllPosts(List<Category> categories) {
        if (Objects.isNull(categories) || categories.isEmpty()) {
            return findAllPosts();
        }

        return categories.stream()
                .map(postRepository::findAllByCategory)
                .flatMap(List::stream)
                .map(PostResponse::of)
                .collect(Collectors.toList());
    }

    public PostResponse findPost(Long id) {
        Post post = postRepository.findById(id).orElseThrow(IllegalArgumentException::new);

        return PostResponse.of(post);
    }

    public Long createPost(PostCreateRequest request, List<MultipartFile> photos, Member member) throws IOException {
        List<Photo> photoUrls = new ArrayList<>();
        for (MultipartFile photo : photos) {
            Photo photoEntity = Photo.builder()
                    .url(uploadService.upload(photo))
                    .build();
            photoUrls.add(photoEntity);
        }

        Post savedPost = postRepository.save(request.toEntity(member, photoUrls));
        return savedPost.getId();
    }
}

File: k-datastore-server/src/main/kotlin/com/db/kdatastoreserver/service/PostService.kt
package com.db.kdatastoreserver.service

import com.db.kdatastoreserver.domain.*
import com.db.kdatastoreserver.domain.repository.PostRepository
import com.db.kdatastoreserver.service.dto.PostCreateRequest
import com.db.kdatastoreserver.service.dto.PostResponse
import org.springframework.stereotype.Service
import java.time.LocalDateTime

@Service
class PostService(
    private val repository: PostRepository,
    private val memberService: MemberService,
    private val s3Service: S3Service
) {

    fun findPost(id: String): PostResponse {
        return repository.findById(id)
            .map { PostResponse(it) }
            .orElseThrow { IllegalArgumentException() }
    }

    fun findAllPost(categories: List<Category>): List<PostResponse> {
        return repository.findAllByCategoryIn(categories)
            .map { PostResponse(it) }
    }

    fun createPost(request: PostCreateRequest, member: Member): String? {
        val photoUrls = request.photos
            .map { s3Service.upload(it) }
            .map { s3Service.retrieve(it) }
            .map { Photo(it) }
            .toList()

        val savedPost = repository.save(
            Post(
                null,
                member,
                request.title,
                request.content,
                photoUrls,
                Location(request.longitude, request.latitude),
                request.price,
                emptyList(),
                request.category,
                Status.SALE,
                LocalDateTime.now(),
                LocalDateTime.now()
            )
        )

        memberService.addWrittenPost(member, savedPost)
        return savedPost._id.toString()
    }
}
